regexfactory/patterns.py

```diff
@@ -1,71 +1,103 @@
-        self.a = a
-        self.z = z
-        regex = f"[{a}-{z}]"
+    def __init__(self, start: str, stop: str):
+        self.start = start
+        self.stop = stop
+        regex = f"[{start}-{stop}]"
         super().__init__(regex)
 
 
 class Set(RegexPattern):
-    def __init__(self, *patterns: Tuple[Union[str, Range]]):
+    def __init__(self, *patterns: ValidPatternType):
         regex = ''
         for p in patterns:
             if isinstance(p, Range):
-                regex += f"{p.a}-{p.z}"
+                regex += f"{p.start}-{p.stop}"
             else:
                 regex += str(p)
         super().__init__(f"[{regex}]")
 
 
 class NotSet(RegexPattern):
-    def __init__(self, *patterns: Tuple[Union[str, Range]]):
+    def __init__(self, *patterns: ValidPatternType):
         regex = ''
         for p in patterns:
             if isinstance(p, Range):
-                regex += f"{p.a}-{p.z}"
+                regex += f"{p.start}-{p.stop}"
             else:
                 regex += str(p)
         super().__init__(f"[^{regex}]")
 
 
 class Amount(RegexPattern):
-    def __init__(
-        self,
-        pattern: Union[str, RegexPattern],
-        i: int,
-        j: int = None,
-        ormore: bool = False
-    ):
+    @overload
+    def __init__(self, pattern: ValidPatternType, repetitions: int, or_more: bool = False):
+        ...
+
+    @overload
+    def __init__(self, pattern: ValidPatternType, minimum: int, maximum: int):
+        ...
+
+    def __init__(self, pattern: ValidPatternType, *args, **kwargs):
+        i, j, or_more = self.parse_init_args(*args, **kwargs)
+
         if j is not None:
             amount = f"{i},{j}"
-        elif ormore:
+        elif or_more:
             amount = f"{i},"
         else:
             amount = f"{i}"
-        super().__init__(pattern + "{" + amount + "}")
+
+        regex = self.get_regex(pattern) + "{" + amount + "}"
+        super().__init__(regex)
+
+    @staticmethod
+    def parse_init_args(*args, **kwargs):
+        s1 = inspect.Signature(
+            parameters=(
+                inspect.Parameter("repetitions", kind=inspect.Parameter.POSITIONAL_OR_KEYWORD),
+                inspect.Parameter("or_more", kind=inspect.Parameter.POSITIONAL_OR_KEYWORD)
+            )
+        )
+        s2 = inspect.Signature(
+            parameters=(
+                inspect.Parameter("minimum", kind=inspect.Parameter.POSITIONAL_OR_KEYWORD),
+                inspect.Parameter("maximum", kind=inspect.Parameter.POSITIONAL_OR_KEYWORD)
+            )
+        )
+        try:
+            bound = s1.bind(*args, **kwargs)
+            i = bound.arguments["repetitions"]
+            j = None
+            or_more = bound.arguments["or_more"]
+        except TypeError:
+            bound = s2.bind(*args, **kwargs)
+            i = bound.arguments["minimum"]
+            j = bound.arguments["maximum"]
+            or_more = False
+
+        return i, j, or_more
```

Comment:
This may be a bit over-the-top for what it's doing, so I will understand if instead of complex logic like this, you'd prefer this code without any overloads and just have less descriptive and less strict function parameters.

Reply:
Another option to avoid this mess would be to convert this into 2 separate class methods instead of having both logic for minimum and maximum or for range with optional `or_more` attribute. This could look like this:

```py
class Amount:
    def __init__(i, j, or_more):
        ...  # The actual logic
        
    @classmethod
    def from_range(cls, minimum: int, maximum: int) -> "Amount":
        return cls(minimum, maximum, False)
        
    @classmethod
    def from_repetitions(cls, repetitions: int, or_more: bool = False) -> "Amount":
        return cls(repetitions, None, or_more)
```

Reply:
Let's go with the overloading direction. It distinguishes the the arguments that can be passed explicitly rather than implying those parameters with less strict default arguments.